autotraining.lua

```diff
@@ -0,0 +1,290 @@
+        if squad then
+            table.insert(squads, squad)
+        else
+            -- setting to nil during iteration is permitted by lua
+            state.training_squads[squad_id] = nil
+        end
+    end
+    return squads
+end
+
+function findNeed(unit)
+    local needs =  unit.status.current_soul.personality.needs
+    for _, need in ipairs(needs) do
+        if need.id == need_id then
+            return need
+        end
+    end
+    return nil
+end
+
+--######
+--Main
+--######
+
+function getByID(id)
+    for _, unit in ipairs(getTrainingCandidates()) do
+        if (unit.hist_figure_id == id) then
+            return unit
+        end
+    end
+
+    return nil
+end
+
+-- Find all training squads
+-- Abort if no squads found
+function checkSquads()
+    local squads = {}
+    for _, squad in ipairs(getTrainingSquads()) do
+        if squad.entity_id == df.global.plotinfo.group_id then
+            local leader = squad.positions[0].occupant
+            if ( leader ~= -1) then
+                table.insert(squads,squad)
+            end
+        end
+    end
+
+    if (#squads == 0) then
+        return nil
+    end
+
+    return squads
+end
+
+function addTraining(unit)
+    if (unit.military.squad_id ~= -1) then
+        for _, squad in ipairs(getTrainingSquads()) do
+            if unit.military.squad_id == squad.id then
+                return true
+            end
+        end
+        return false
+    end
+    for _, squad in ipairs(getTrainingSquads()) do
+        for i=1,9,1   do
+            if ( squad.positions[i].occupant  == -1 ) then
+                return dfhack.military.addToSquad(unit.id,squad.id,i)
+            end
+        end
+    end
+
+    return false
+end
+
+function removeTraining(unit)
+    return dfhack.military.removeFromSquad(unit.id)
+end
+
+function removeAll()
+    if ( state.training_squads == nil) then return end
+    for _, squad in ipairs(getTrainingSquads()) do
+        for i=1,9,1 do
+            local dwarf = getByID(squad.positions[i].occupant)
+            if (dwarf ~= nil) then
+                removeTraining(dwarf)
+            end
+        end
+    end
+end
+
+
+function check()
+    local squads = checkSquads()
+    local intraining_count = 0
+    local inque_count = 0
+    if ( squads == nil) then return end
+    for _, unit in ipairs(getTrainingCandidates()) do
+        local need = findNeed(unit)
+        if ( need  ~= nil ) then
+            if ( need.focus_level  < state.threshold ) then
+                local bol = addTraining(unit)
+                if ( bol ) then
+                    intraining_count = intraining_count +1
+                else
+                    inque_count = inque_count +1
+                end
+            else
+                removeTraining(unit)
+            end
+        end
+    end
+
+    dfhack.println(GLOBAL_KEY  .. " | IGN: " .. ignore_count .. " TRAIN: " .. intraining_count .. " QUE: " ..inque_count )
```

Comment:
This should probably use full words rather than abbreviations and be the only output during regular operation. Similarly, enabling and disabling should not generate output. `enable` follows the unix philosophy: no output means success.